### ISSUE TITLE
Moving asynctasks to the background scheduler caused deadlocks

### ISSUE THREAD

**Issue by charlesmunger** [robolectric/robolectric#2115]
#1043 is back, this time with vanilla AsyncTasks.

I think @kriegfrj's recent changes (3f4afb648e4140db48a7c12748c8eeadeb985fba and friends) triggered this. 
```
"AsyncTask #1" Id=17 BLOCKED on org.robolectric.util.Scheduler@1c064dd4 owned by "main" Id=1
	at org.robolectric.util.Scheduler.getCurrentTime(Scheduler.java:114)
	-  blocked on org.robolectric.util.Scheduler@1c064dd4
	at org.robolectric.shadows.ShadowSystemClock.currentTimeNanos(ShadowSystemClock.java:92)
	at org.robolectric.shadows.ShadowSystemClock.currentTimeMillis(ShadowSystemClock.java:108)
	at sun.reflect.GeneratedMethodAccessor2.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.robolectric.util.ReflectionHelpers.callStaticMethod(ReflectionHelpers.java:226)
	at org.robolectric.internal.bytecode.ShadowWrangler$6.call(ShadowWrangler.java:279)
	at org.robolectric.internal.bytecode.ShadowWrangler.intercept(ShadowWrangler.java:228)
	at org.robolectric.internal.bytecode.RobolectricInternals.intercept(RobolectricInternals.java:33)
	at android.database.sqlite.SQLiteConnection$OperationLog.$$robo$$beginOperation(SQLiteConnection.java:1311)
	-  locked [Landroid.database.sqlite.SQLiteConnection$Operation;@5904c890
	at android.database.sqlite.SQLiteConnection$OperationLog.beginOperation(SQLiteConnection.java)
	at android.database.sqlite.SQLiteConnection.$$robo$$prepare(SQLiteConnection.java:498)
	at android.database.sqlite.SQLiteConnection.prepare(SQLiteConnection.java)
	at android.database.sqlite.SQLiteSession.$$robo$$prepare(SQLiteSession.java:588)
	at android.database.sqlite.SQLiteSession.prepare(SQLiteSession.java)
	at android.database.sqlite.SQLiteProgram.$$robo$$__constructor__(SQLiteProgram.java:58)
	at android.database.sqlite.SQLiteProgram.<init>(SQLiteProgram.java)
	at android.database.sqlite.SQLiteStatement.<init>(SQLiteStatement.java)
	at android.database.sqlite.SQLiteDatabase.$$robo$$updateWithOnConflict(SQLiteDatabase.java:1572)
	at android.database.sqlite.SQLiteDatabase.updateWithOnConflict(SQLiteDatabase.java)
	at android.database.sqlite.SQLiteDatabase.$$robo$$update(SQLiteDatabase.java:1520)
	at android.database.sqlite.SQLiteDatabase.update(SQLiteDatabase.java)
	at redacted
	at org.robolectric.shadows.ShadowContentResolver.update(ShadowContentResolver.java:141)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.robolectric.internal.bytecode.ShadowWrangler$ShadowMethodPlan.run(ShadowWrangler.java:438)
	at android.content.ContentResolver.update(ContentResolver.java)
	at redacted
	at redacted
	at android.os.AsyncTask$2.$$robo$$call(AsyncTask.java:288)
	at android.os.AsyncTask$2.call(AsyncTask.java)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.robolectric.shadows.ShadowAsyncTask$FutureTaskWrapper.run(ShadowAsyncTask.java:140)
	at android.os.AsyncTask$SerialExecutor$1.$$robo$$run(AsyncTask.java:231)
	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```
and:
```
"main" Id=1 WAITING on java.util.concurrent.CountDownLatch$Sync@731198f0
	at sun.misc.Unsafe.park(Native Method)
	-  waiting on java.util.concurrent.CountDownLatch$Sync@731198f0
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
	at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
	at org.robolectric.shadows.ShadowAsyncTask$1.run(ShadowAsyncTask.java:191)
	at org.robolectric.util.Scheduler.runOrQueueRunnable(Scheduler.java:364)
	at org.robolectric.util.Scheduler.postDelayed(Scheduler.java:179)
	-  locked org.robolectric.util.Scheduler@62c0bd3d
	at org.robolectric.util.Scheduler.postDelayed(Scheduler.java:164)
	-  locked org.robolectric.util.Scheduler@62c0bd3d
	at org.robolectric.util.Scheduler.post(Scheduler.java:152)
	-  locked org.robolectric.util.Scheduler@62c0bd3d
	at org.robolectric.shadows.ShadowAsyncTask.executeOnExecutor(ShadowAsyncTask.java:187)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.robolectric.internal.bytecode.ShadowWrangler$ShadowMethodPlan.run(ShadowWrangler.java:438)
	at android.os.AsyncTask.executeOnExecutor(AsyncTask.java)
	at android.os.AsyncTask.$$robo$$execute(AsyncTask.java:535)
	at android.os.AsyncTask.execute(AsyncTask.java)
	at redacted
	at android.support.v4.app.Fragment.$$robo$$onAttach(Fragment.java:1152)
	at android.support.v4.app.Fragment.onAttach(Fragment.java)
	at android.support.v4.app.FragmentManagerImpl.$$robo$$moveToState(FragmentManager.java:968)
	at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java)
	at android.support.v4.app.FragmentManagerImpl.$$robo$$moveToState(FragmentManager.java:1197)
	at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java)
	at android.support.v4.app.BackStackRecord.$$robo$$run(BackStackRecord.java:738)
	at android.support.v4.app.BackStackRecord.run(BackStackRecord.java)
	at android.support.v4.app.FragmentManagerImpl.$$robo$$execPendingActions(FragmentManager.java:1562)
	at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java)
	at android.support.v4.app.FragmentManagerImpl$1.$$robo$$run(FragmentManager.java:483)
	at android.support.v4.app.FragmentManagerImpl$1.run(FragmentManager.java)
	at android.os.Handler.$$robo$$handleCallback(Handler.java:733)
	at android.os.Handler.handleCallback(Handler.java)
	at android.os.Handler.$$robo$$dispatchMessage(Handler.java:95)
	at android.os.Handler.dispatchMessage(Handler.java)
	at org.robolectric.shadows.ShadowMessageQueue.dispatchMessage(ShadowMessageQueue.java:140)
	at org.robolectric.shadows.ShadowMessageQueue.access$100(ShadowMessageQueue.java:30)
	at org.robolectric.shadows.ShadowMessageQueue$1.run(ShadowMessageQueue.java:105)
	-  locked android.os.MessageQueue@496be4df
	at org.robolectric.util.Scheduler.runOrQueueRunnable(Scheduler.java:364)
	at org.robolectric.util.Scheduler.postDelayed(Scheduler.java:179)
	-  locked org.robolectric.util.Scheduler@1c064dd4
	at org.robolectric.util.Scheduler.postDelayed(Scheduler.java:164)
	-  locked org.robolectric.util.Scheduler@1c064dd4
	at org.robolectric.shadows.ShadowMessageQueue.enqueueMessage(ShadowMessageQueue.java:126)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.robolectric.internal.bytecode.ShadowWrangler$ShadowMethodPlan.run(ShadowWrangler.java:438)
	at android.os.MessageQueue.enqueueMessage(MessageQueue.java)
	at android.os.Handler.$$robo$$enqueueMessage(Handler.java:626)
	at android.os.Handler.enqueueMessage(Handler.java)
	at android.os.Handler.$$robo$$sendMessageAtTime(Handler.java:595)
	at android.os.Handler.sendMessageAtTime(Handler.java)
	at android.os.Handler.$$robo$$sendMessageDelayed(Handler.java:566)
	at android.os.Handler.sendMessageDelayed(Handler.java)
	at android.os.Handler.$$robo$$post(Handler.java:326)
	at android.os.Handler.post(Handler.java)
	at android.support.v4.app.FragmentManagerImpl.$$robo$$enqueueAction(FragmentManager.java:1468)
	-  locked android.support.v4.app.FragmentManagerImpl@78672a03
	at android.support.v4.app.FragmentManagerImpl.enqueueAction(FragmentManager.java)
	at android.support.v4.app.BackStackRecord.$$robo$$commitInternal(BackStackRecord.java:634)
	at android.support.v4.app.BackStackRecord.commitInternal(BackStackRecord.java)
	at android.support.v4.app.BackStackRecord.$$robo$$commit(BackStackRecord.java:613)
	at android.support.v4.app.BackStackRecord.commit(BackStackRecord.java)
	at org.robolectric.shadows.support.v4.SupportFragmentTestUtil.startVisibleFragment(SupportFragmentTestUtil.java:32)
	at redacted
	at redacted
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:49)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:551)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.robolectric.RobolectricTestRunner$2.evaluate(RobolectricTestRunner.java:258)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:188)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:54)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:240)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:65)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:238)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:55)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:231)
	at org.robolectric.RobolectricTestRunner$1.evaluate(RobolectricTestRunner.java:152)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:316)
	at redacted
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:138)
	at redacted
	at redacted
	at redacted
```

**Comment by charlesmunger** [comment]
I propose reverting those changes until we have an answer for the deadlocks.

**Comment by erd** [comment]
I'd like to give @kriegfrj a chance to take a look.

**Comment by kriegfrj** [comment]
I'll try and have a look tonight. In the meantime if @charlesmunger can post the code that caused the deadlock (or a redacted version of it), that would be appreciated.

**Comment by kriegfrj** [comment]
Had a bit of a look at those stack traces and I can see where the deadlock is happening. I'm not near a computer but if you get a chance to look at it before I do @charlesmunger, try modifying Scheduler to make the currentTime instance variable volatile and change getCurrentTime() so that it is not synchronized.

**Comment by kriegfrj** [comment]
I think that #2116 should fix this particular deadlock.

Note also that #2095 also has the potential to expose real deadlocks in app code itself which hitherto would not have been detected by Robolectric (due to the fact that bg tasks actually ran on the main thread). Consider this example (perhaps slightly contrived, but will illustrate the point):

````
public void test() {
  // Assume fg and bg schedulers are both unpaused.
  Object lock = new Object();
  AsyncTask<?,?,?> t = new AsyncTask<String,String,String>() {
    @Override
    public String doInBackground(String... params) {
      synchronized (lock) {
        doStuff();
      }
    }
  };
  synchronized (lock) {
    t.execute("param").get();
  }
}
````
This code would happily execute in Robolectric prior to #2095 because a thread is allowed to renter a lock that it already holds, but in a real Android situation this would deadlock when doInBackground() tries to acquire the lock. Post #2095 Robolectric will deadlock too in this scenario.

**Comment by charlesmunger** [comment]
I agree that executing asynctasks on multiple threads is the correct way for robolectric moving forward, but I also don't think we should do it until we have our locking story straight. Fundamentally, as long as we're calling user code while holding a scheduler lock, we're going to have deadlocks. Consider the simple case of posting to the UI thread:
```
public void test() {
  // Assume fg and bg schedulers are both unpaused.
  Object lock = new Object();
  AsyncTask<?,?,?> t = new AsyncTask<String,String,String>() {
    @Override
    public String doInBackground(String... params) {
       postOnUiThread(someRunnable);
    }
  };
  postOnUiThread(new Runnable () {
    public void run() {
        t.execute("param").get();
    }
  });
}
```

This also deadlocks. Any code that posts events from inside a Scheduler is going to cause deadlock.

**Comment by kriegfrj** [comment]
This is a good point and echoes something that @perlow noted in #1043:  "You don't call arbitrary code while holding the lock. That's right out of the textbook of how to cause deadlocks." Scheduler is definitely over-heavy on the locking and #2116 is only part of the story.

**Comment by caltseng** [comment]
I am still deadlocking on this. Looking into the code, it looks like getCurrentTime() is still synchronized .

> "pool-5-thread-1":
	at org.robolectric.util.Scheduler.getCurrentTime(Scheduler.java:93)
	- waiting to lock <0x00000006c3394018> (a org.robolectric.util.Scheduler)
	at org.robolectric.shadows.ShadowSystemClock.now(ShadowSystemClock.java:25)
	at org.robolectric.shadows.ShadowSystemClock.uptimeMillis(ShadowSystemClock.java:54)
	at java.lang.invoke.LambdaForm$DMH/638169719.invokeStatic__J(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/211992147.delegate(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/616412281.guard(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/2118096382.linkToCallSite(LambdaForm$MH)
	at android.os.SystemClock.uptimeMillis(SystemClock.java)
	at android.database.sqlite.SQLiteConnectionPool.$$robo$$waitForConnection(SQLiteConnectionPool.java:621)
	- locked <0x00000007922334e0> (a java.lang.Object)
	at java.lang.invoke.LambdaForm$DMH/1800500978.invokeSpecial_LLIL_L(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/1139851985.guardWithCatch(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/799755007.delegate(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1145882513.guard(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/475816303.linkToCallSite(LambdaForm$MH)
	at android.database.sqlite.SQLiteConnectionPool.waitForConnection(SQLiteConnectionPool.java)
	at android.database.sqlite.SQLiteConnectionPool.$$robo$$acquireConnection(SQLiteConnectionPool.java:348)
	at java.lang.invoke.LambdaForm$DMH/1800500978.invokeSpecial_LLIL_L(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/1139851985.guardWithCatch(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/799755007.delegate(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1145882513.guard(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/475816303.linkToCallSite(LambdaForm$MH)
	at android.database.sqlite.SQLiteConnectionPool.acquireConnection(SQLiteConnectionPool.java)
	at android.database.sqlite.SQLiteSession.$$robo$$acquireConnection(SQLiteSession.java:894)
	at java.lang.invoke.LambdaForm$DMH/1850137076.invokeSpecial_LLIL_V(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/2027371291.guardWithCatch(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/566641954.delegate(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/518739992.guard(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1155521880.linkToCallSite(LambdaForm$MH)
	at android.database.sqlite.SQLiteSession.acquireConnection(SQLiteSession.java)
	at android.database.sqlite.SQLiteSession.$$robo$$executeForCursorWindow(SQLiteSession.java:834)
	at java.lang.invoke.LambdaForm$DMH/171011947.invokeSpecial_L4I4L_I(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/2061865206.guardWithCatch(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1645578641.delegate(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/710589756.guard(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/621373425.linkToCallSite(LambdaForm$MH)
	at android.database.sqlite.SQLiteSession.executeForCursorWindow(SQLiteSession.java)
	at android.database.sqlite.SQLiteQuery.$$robo$$fillWindow(SQLiteQuery.java:62)
	at java.lang.invoke.LambdaForm$DMH/831654622.invokeSpecial_LLI3_I(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/412416671.guardWithCatch(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1332439166.delegate(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/869413070.guard(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/904351240.linkToCallSite(LambdaForm$MH)
	at android.database.sqlite.SQLiteQuery.fillWindow(SQLiteQuery.java)
	at android.database.sqlite.SQLiteCursor.$$robo$$fillWindow(SQLiteCursor.java:143)
	at java.lang.invoke.LambdaForm$DMH/144040807.invokeSpecial_LI_V(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/970419381.guardWithCatch(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/318794136.delegate(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1272123492.guard(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/238816832.linkToCallSite(LambdaForm$MH)
	at android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java)
	at android.database.sqlite.SQLiteCursor.$$robo$$getCount(SQLiteCursor.java:132)
	at java.lang.invoke.LambdaForm$DMH/13803304.invokeSpecial_L_I(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/1422273905.guardWithCatch(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/890545344.delegate(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/264394929.guard(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1878413714.linkToCallSite(LambdaForm$MH)
	at android.database.sqlite.SQLiteCursor.getCount(SQLiteCursor.java)
	at org.greenrobot.greendao.AbstractDao.loadAllFromCursor(AbstractDao.java:379)
	at org.greenrobot.greendao.AbstractDao.loadAllAndCloseCursor(AbstractDao.java:194)
	at org.greenrobot.greendao.InternalQueryDaoAccess.loadAllAndCloseCursor(InternalQueryDaoAccess.java:21)
	at org.greenrobot.greendao.query.Query.list(Query.java:77)
	at org.greenrobot.greendao.async.AsyncOperationExecutor.executeOperation(AsyncOperationExecutor.java:311)
	at org.greenrobot.greendao.async.AsyncOperationExecutor.executeOperationAndPostCompleted(AsyncOperationExecutor.java:259)
	at org.greenrobot.greendao.async.AsyncOperationExecutor.run(AsyncOperationExecutor.java:167)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
"main":
	at android.database.sqlite.SQLiteConnectionPool.$$robo$$releaseConnection(SQLiteConnectionPool.java:364)
	- waiting to lock <0x00000007922334e0> (a java.lang.Object)
	at java.lang.invoke.LambdaForm$DMH/2103569237.invokeSpecial_LL_V(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/628392976.guardWithCatch(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1046665075.delegate(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/694069884.guard(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/957923719.linkToCallSite(LambdaForm$MH)
	at android.database.sqlite.SQLiteConnectionPool.releaseConnection(SQLiteConnectionPool.java)
	at android.database.sqlite.SQLiteSession.$$robo$$releaseConnection(SQLiteSession.java:906)
	at java.lang.invoke.LambdaForm$DMH/1095352419.invokeSpecial_L_V(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/573200870.guardWithCatch(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/722951168.delegate(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1670313965.guard(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1275035040.linkToCallSite(LambdaForm$MH)
	at android.database.sqlite.SQLiteSession.releaseConnection(SQLiteSession.java)
	at android.database.sqlite.SQLiteSession.$$robo$$endTransactionUnchecked(SQLiteSession.java:442)
	at java.lang.invoke.LambdaForm$DMH/1241569743.invokeSpecial_LLI_V(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/864248990.guardWithCatch(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1345401730.delegate(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1483145520.guard(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1961129028.linkToCallSite(LambdaForm$MH)
	at android.database.sqlite.SQLiteSession.endTransactionUnchecked(SQLiteSession.java)
	at android.database.sqlite.SQLiteSession.$$robo$$endTransaction(SQLiteSession.java:401)
	at java.lang.invoke.LambdaForm$DMH/2103569237.invokeSpecial_LL_V(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/628392976.guardWithCatch(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1046665075.delegate(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/694069884.guard(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/957923719.linkToCallSite(LambdaForm$MH)
	at android.database.sqlite.SQLiteSession.endTransaction(SQLiteSession.java)
	at android.database.sqlite.SQLiteDatabase.$$robo$$endTransaction(SQLiteDatabase.java:522)
	at java.lang.invoke.LambdaForm$DMH/1095352419.invokeSpecial_L_V(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/573200870.guardWithCatch(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/722951168.delegate(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1670313965.guard(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1275035040.linkToCallSite(LambdaForm$MH)
	at android.database.sqlite.SQLiteDatabase.endTransaction(SQLiteDatabase.java)
	at org.greenrobot.greendao.database.StandardDatabase.endTransaction(StandardDatabase.java:46)
	at org.greenrobot.greendao.AbstractDao.executeInsert(AbstractDao.java:345)
	at org.greenrobot.greendao.AbstractDao.insertOrReplace(AbstractDao.java:331)
	at ols.microsoft.com.shiftr.service.ShiftrDataService.updateSyncDataAndNotify(ShiftrDataService.java:7234)
	at ols.microsoft.com.shiftr.service.ShiftrDataService.access$10200(ShiftrDataService.java:233)
	at ols.microsoft.com.shiftr.service.ShiftrDataService$142$2.onSuccess(ShiftrDataService.java:7548)
	at ols.microsoft.com.shiftr.service.ShiftrDataService$31.onOperationCompleted(ShiftrDataService.java:1713)
	at ols.microsoft.com.shiftr.service.ShiftrDataService.onAsyncOperationCompleted(ShiftrDataService.java:3265)
	at org.greenrobot.greendao.async.AsyncOperationExecutor.handleMessage(AsyncOperationExecutor.java:371)
	at android.os.Handler.$$robo$$dispatchMessage(Handler.java:98)
	at java.lang.invoke.LambdaForm$DMH/2103569237.invokeSpecial_LL_V(LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/628392976.guardWithCatch(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/1046665075.delegate(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/694069884.guard(LambdaForm$MH)
	at java.lang.invoke.LambdaForm$MH/957923719.linkToCallSite(LambdaForm$MH)
	at android.os.Handler.dispatchMessage(Handler.java)
	at org.robolectric.shadows.ShadowMessageQueue.dispatchMessage(ShadowMessageQueue.java:144)
	at org.robolectric.shadows.ShadowMessageQueue.access$100(ShadowMessageQueue.java:30)
	at org.robolectric.shadows.ShadowMessageQueue$1.run(ShadowMessageQueue.java:105)
	- locked <0x00000006c44f7bd0> (a android.os.MessageQueue)
	at org.robolectric.util.Scheduler$ScheduledRunnable.run(Scheduler.java:354)
	at org.robolectric.util.Scheduler.runOneTask(Scheduler.java:256)
	- locked <0x00000006c3394018> (a org.robolectric.util.Scheduler)
	at org.robolectric.util.Scheduler.advanceTo(Scheduler.java:237)
	- locked <0x00000006c3394018> (a org.robolectric.util.Scheduler)
	at org.robolectric.util.Scheduler.advanceBy(Scheduler.java:220)
	- locked <0x00000006c3394018> (a org.robolectric.util.Scheduler)
	at org.robolectric.shadows.ShadowLooper.idle(ShadowLooper.java:233)
	at org.robolectric.shadows.ShadowLooper.idle(ShadowLooper.java:214)
...

**Comment by kriegfrj** [comment]
Sorry @caltseng, I had a fix for this but it stalled in the face of various life commitments and other features that were tightly related to it. Maybe soon? Can't promise anything though...

**Comment by jongerrish** [comment]
Should be fixed by #2982 - please reopen if not.
